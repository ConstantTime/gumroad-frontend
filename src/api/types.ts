export interface IFetchDataParams {
  endpoint: string;
  method: EHttpMethod;
  body?: object;
  payloadToQueryParam?: boolean;
}

export enum EHttpMethod {
  GET = "GET",
  POST = "POST",
  PUT = "PUT",
  PATCH = "PATCH",
  DELETE = "DELETE",
}

export interface IInitOptions {
  method: EHttpMethod;
  body?: {};
}
