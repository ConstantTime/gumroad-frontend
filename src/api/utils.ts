import { merge } from "timm";
import { EHttpMethod, IFetchDataParams, IInitOptions } from "./types";

/**
 * Construct initOptions for the Request object used by fetch()
 * @param method HTTP method to use for the API request
 * @param body Request payload (URL parameters or request body)
 */
export const getInitOptions = ({ method, body }: IInitOptions): RequestInit => {
  const initOptions: RequestInit = {
    mode: "cors",
    method,
  };

  if (method === EHttpMethod.GET) {
    return initOptions;
  }

  if (body) {
    return merge(initOptions, {
      body: JSON.stringify(body),
      headers: {
        "Content-Type": "application/json",
      },
    });
  }
  return merge(initOptions, {
    body: null,
  });
};

/**
 * Add params as query parameters to the URL
 * @param url URL without query parameters
 * @param params Object to encode as query parameters
 */
export const appendUrlParameters = (url: string, params: any) => {
  if (typeof params !== "object") {
    return url;
  }

  const paramsKeys = Object.keys(params);

  if (!paramsKeys.length || !url) {
    return url;
  }

  // Create an array of `key=value` strings
  const paramsArr = paramsKeys
    // Filter all truthy or 0 value params
    .filter((key) => params[key] || params[key] === 0)
    .map((key) => {
      const pVal = params[key];
      let encodedVal = "";

      if (typeof pVal === "object") {
        encodedVal = encodeURIComponent(JSON.stringify(pVal));
      } else {
        encodedVal = encodeURIComponent(pVal);
      }

      return `${key}=${encodedVal}`;
    });

  if (!paramsArr.length) {
    return url;
  }

  // Join all params with `&` and add them to the URL
  return `${url}?${paramsArr.join("&")}`;
};

export const callApi = async <T = {}>({
  endpoint,
  method,
  body,
  payloadToQueryParam = false,
}: IFetchDataParams) => {
  try {
    let request: Request;
    let url: string = endpoint;
    if (method === EHttpMethod.GET || payloadToQueryParam) {
      url = appendUrlParameters(url, body);
      request = new Request(url, getInitOptions({ method }));
    } else {
      request = new Request(url, getInitOptions({ method, body }));
    }
    const response = await fetch(request);

    return response;
  } catch (e) {
    // No op
  }
};
