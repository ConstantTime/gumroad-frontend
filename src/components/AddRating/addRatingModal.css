.add-rating-modal {
  display: flex;
  flex-direction: column;
  width: 40%;
  margin: 0 auto;
  margin-top: 200px;
  background-color: white;
  align-items: flex-start;
  padding: 30px;
  min-height: 300px;
  justify-content: space-around;
}

.add-rating-modal .MuiButtonBase-root {
  margin-top: 30px;
}
