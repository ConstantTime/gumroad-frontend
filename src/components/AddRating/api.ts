import axios from "axios";
import { ApiEndpoints, SERVER_URL } from "../../constants/Constants";

export const addRatingAndReview = (id: any, rating: number, review: String) => {
  axios
    .request({
      url: SERVER_URL + ApiEndpoints.ADD_RATING_AND_REVIEW(id.id as String),
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      data: JSON.stringify({
        rating: rating,
        review: review,
      }),
    })
    .catch((response) => {
      console.log("Error while adding rating:", response);
    });
};
