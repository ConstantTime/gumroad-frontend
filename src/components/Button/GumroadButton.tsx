import Button from "@material-ui/core/Button";
import React from "react";
import { IGumroadButton } from "./types";

const GumroadButton: React.FC<IGumroadButton> = ({ onClick, children }) => {
  return (
    <Button
      style={{ backgroundColor: "#FFFFFF", color: "#000000" }}
      size="medium"
      variant="contained"
      onClick={onClick}
    >
      {children}
    </Button>
  );
};
export default GumroadButton;
