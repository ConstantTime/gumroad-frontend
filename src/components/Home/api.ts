import axios from "axios";
import { IProduct } from "../../api/responses/types";
import { ApiEndpoints, SERVER_URL } from "../../constants/Constants";

export const fetchProducts = (setProducts: (data: IProduct[]) => void) => {
  axios
    .get<IProduct[]>(SERVER_URL + ApiEndpoints.FETCH_PRODUCTS, {
      method: "GET",
    })
    .then((response) => {
      setProducts(response.data);
    })
    .catch((response) => {
      console.log("Error:", response);
    });
};
