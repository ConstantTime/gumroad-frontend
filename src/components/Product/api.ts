import axios from "axios";
import { IProduct, IRating } from "../../api/responses/types";
import { ApiEndpoints, SERVER_URL } from "../../constants/Constants";
import "./product.css";

export const fetchProductDetails = (
  id: any,
  setProduct: (data: IProduct | undefined) => void
) => {
  axios
    .get<IProduct>(
      SERVER_URL + ApiEndpoints.FETCH_PRODUCT_DETAILS(id.id as String),
      {
        method: "GET",
      }
    )
    .then((response) => {
      setProduct(response.data);
    })
    .catch((response) => {
      console.log("Error while fetching product details:", response);
    });
};

export const fetchTop3Ratings = (
  id: any,
  setRatings: (data: IRating[]) => void
) => {
  axios
    .get<IRating[]>(
      SERVER_URL + ApiEndpoints.FETCH_TOP3_RATINGS(id.id as String),
      {
        method: "GET",
      }
    )
    .then((response) => {
      setRatings(response.data);
    })
    .catch((response) => {
      console.log("Error while fetching top 3 ratings:", response);
    });
};
