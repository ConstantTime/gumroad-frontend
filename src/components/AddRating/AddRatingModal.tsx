import TextField from "@material-ui/core/TextField";
import Typography from "@material-ui/core/Typography";
import StarBorderIcon from "@material-ui/icons/StarBorder";
import Rating from "@material-ui/lab/Rating";
import axios from "axios";
import React, { useState } from "react";
import { useParams } from "react-router-dom";
import { ApiEndpoints, SERVER_URL } from "../../constants/Constants";
import GumroadButton from "../Button/GumroadButton";
import "./addRatingModal.css";

const DEFAULT_VALUE = 3;
const AddRatingModal: React.FC = () => {
  const id = useParams<any>();
  const [rating, setRating] = useState(DEFAULT_VALUE);
  const [review, setReview] = useState("");
  const handleClick = () => {
    console.log(rating, review);
    axios
      .request({
        url: SERVER_URL + ApiEndpoints.ADD_RATING_AND_REVIEW(id.id as String),
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        data: JSON.stringify({
          rating: rating,
          review: review,
        }),
      })
      .catch((response) => {
        console.log("Error while adding rating:", response);
      });
  };

  const handleChangeInReview = (
    e: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
  ) => {
    setReview(e.target.value);
  };

  return (
    <div className="add-rating-modal">
      <Typography variant="h4" component="h2" align="center">
        What's your rating?
      </Typography>

      <Typography variant="h6" component="h2" align="center">
        Rating
      </Typography>

      <Rating
        name="product-rating"
        defaultValue={DEFAULT_VALUE}
        precision={0.5}
        onChange={(_, newValue) => {
          newValue != null && setRating(newValue);
        }}
        emptyIcon={<StarBorderIcon fontSize="inherit" />}
      />

      <Typography variant="h6" component="h2" align="center">
        Review
      </Typography>

      <TextField
        id="standard-basic"
        label="Start typing"
        onChange={(e) => handleChangeInReview(e)}
      />

      <GumroadButton onClick={handleClick}>Submit Review</GumroadButton>
    </div>
  );
};

export default AddRatingModal;
