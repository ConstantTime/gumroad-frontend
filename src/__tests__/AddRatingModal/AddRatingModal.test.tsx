import { shallow } from "enzyme";
import AddRatingModal from "../../components/AddRating/AddRatingModal";
import * as apiUtils from "../../components/AddRating/api";
import GumroadButton from "../../components/Button/GumroadButton";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: {
      id: "test",
    },
  }),
}));

describe("Add rating modal", () => {
  it("should render the component", () => {
    const wrapper = shallow(<AddRatingModal closeModal={jest.fn()} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should add rating and review and close the modal on submitting review", () => {
    const handleClose = jest.fn();
    const mockAddRatingAndReview = jest.spyOn(apiUtils, "addRatingAndReview");

    const wrapper = shallow(<AddRatingModal closeModal={handleClose} />);
    const button = wrapper.find(GumroadButton);
    expect(button.exists()).toBeTruthy();
    button.simulate("click");

    expect(handleClose).toHaveBeenCalled();
    expect(mockAddRatingAndReview).toHaveBeenCalled();
  });
});
