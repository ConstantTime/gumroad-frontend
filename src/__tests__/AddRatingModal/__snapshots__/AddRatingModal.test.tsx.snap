// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add rating modal should render the component 1`] = `
<div
  className="add-rating-modal"
>
  <WithStyles(ForwardRef(Typography))
    align="center"
    component="h2"
    variant="h4"
  >
    What's your rating?
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Typography))
    align="center"
    component="h2"
    variant="h6"
  >
    Rating
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(Rating))
    defaultValue={3}
    emptyIcon={
      <Memo
        fontSize="inherit"
      />
    }
    name="product-rating"
    onChange={[Function]}
    precision={0.5}
  />
  <WithStyles(ForwardRef(Typography))
    align="center"
    component="h2"
    variant="h6"
  >
    Review
  </WithStyles(ForwardRef(Typography))>
  <WithStyles(ForwardRef(TextField))
    id="standard-basic"
    label="Start typing"
    onChange={[Function]}
  />
  <GumroadButton
    onClick={[Function]}
  >
    Submit Review
  </GumroadButton>
</div>
`;
